### PR TITLE
Stop tracking skb in the end of its lifetime

### PR DESCRIPTION
**Pull request by jschwinger233**
`--filter-track-skb` has a problem of mismatching skbs with same pointer address. This PR tries to fix it by stopping tracking after an skb ends its lifetime.

In general, an skb has 3 possible endings:
1. consumed locally
2. dropped locally
3. transferred via a netdev

All the three endings hit `kfree_skbmem` eventually, so we delete tracked skb pointer address from bpf map at that time.

Fixes: https://github.com/cilium/pwru/pull/194

Signed-off-by: Zhichuan Liang <gray.liang@isovalent.com>

### CODE REVIEW

**Comment by jschwinger233** [comment]
The outstanding issue is, I can't come up with a good flag name.

In the 3rd commit, I added a flag `--filter-track-skb-across-namespaces` to control if we see `eth_type_trans` as a lifetime termination. 
For a containerized cluster such as kind cluster, we don't want to stop tracking on `eth_type_trans`; for bare metal node, or in the case we want to focus on a specific netns, we definitely want to stop at `eth_type_trans`. 

But I don't think `--filter-track-skb-across-namespaces` is a good name. Do you have some idea? @brb

File: main.go [hunk 1]
```diff
@@ -187,6 +191,27 @@ func main() {
 	log.Printf("Attaching kprobes (via %s)...\n", msg)
 	ignored := 0
 	bar := pb.StartNew(len(funcs))
+
+	if flags.FilterTrackSkb {
+		skbLifetimeEndAt := []string{"kfree_skbmem"}
+		if !flags.FilterTrackSkbAcrossNamespaces {
+			skbLifetimeEndAt = append(skbLifetimeEndAt, "eth_type_trans")
```
Comment (jschwinger233) [inline main.go:198]: I don't know how to do it flexibly without hard coding `eth_type_trans` and `kfree_skbmem` - -

**Comment by brb** [comment]
> But I don't think --filter-track-skb-across-namespaces is a good name. Do you have some idea? @brb

Tough one :sweat_smile: What's about `--filter-stop-track-skb-after-emit`?

**Comment by jschwinger233** [comment]
Marking `kfree_skbmem` as skb lifetime termination looks fine.

As for `eth_type_trans`, it definitely is a bad choice.

Below is a typical pwru output for an egress traffic inside a netns:

```
0xffff97bfe0bb54e8      0           [curl]             ip_local_out netns=4026535338 mark=0x0 ifindex=0 proto=0 mtu=0 len=60 10.0.1.45:54816->10.0.0.99:80(tcp)
0xffff97bfe0bb54e8      0           [curl]           __ip_local_out netns=4026535338 mark=0x0 ifindex=0 proto=0 mtu=0 len=60 10.0.1.45:54816->10.0.0.99:80(tcp)
0xffff97bfe0bb54e8      0           [curl]                ip_output netns=4026535338 mark=0x0 ifindex=0 proto=8 mtu=0 len=60 10.0.1.45:54816->10.0.0.99:80(tcp)
0xffff97bfe0bb54e8      0           [curl]             nf_hook_slow netns=4026535338 mark=0x0 ifindex=13 proto=8 mtu=1500 len=60 10.0.1.45:54816->10.0.0.99:80(tcp)
0xffff97bfe0bb54e8      0           [curl]    apparmor_ip_postroute netns=4026535338 mark=0x0 ifindex=13 proto=8 mtu=1500 len=60 10.0.1.45:54816->10.0.0.99:80(tcp)
0xffff97bfe0bb54e8      0           [curl]         ip_finish_output netns=4026535338 mark=0x0 ifindex=13 proto=8 mtu=1500 len=60 10.0.1.45:54816->10.0.0.99:80(tcp)
0xffff97bfe0bb54e8      0           [curl]       __ip_finish_output netns=4026535338 mark=0x0 ifindex=13 proto=8 mtu=1500 len=60 10.0.1.45:54816->10.0.0.99:80(tcp)
0xffff97bfe0bb54e8      0           [curl]        ip_finish_output2 netns=4026535338 mark=0x0 ifindex=13 proto=8 mtu=1500 len=60 10.0.1.45:54816->10.0.0.99:80(tcp)
0xffff97bfe0bb54e8      0           [curl]     neigh_resolve_output netns=4026535338 mark=0x0 ifindex=13 proto=8 mtu=1500 len=60 10.0.1.45:54816->10.0.0.99:80(tcp)
0xffff97bfe0bb54e8      0           [curl]       __neigh_event_send netns=4026535338 mark=0x0 ifindex=13 proto=8 mtu=1500 len=60 10.0.1.45:54816->10.0.0.99:80(tcp)
0xffff97bfe0bb54e8      0           [curl]               eth_header netns=4026535338 mark=0x0 ifindex=13 proto=8 mtu=1500 len=60 10.0.1.45:54816->10.0.0.99:80(tcp)
0xffff97bfe0bb54e8      0           [curl]                 skb_push netns=4026535338 mark=0x0 ifindex=13 proto=8 mtu=1500 len=60 10.0.1.45:54816->10.0.0.99:80(tcp)
0xffff97bfe0bb54e8      0           [curl]         __dev_queue_xmit netns=4026535338 mark=0x0 ifindex=13 proto=8 mtu=1500 len=74 10.0.1.45:54816->10.0.0.99:80(tcp)
0xffff97bfe0bb54e8      0           [curl]       qdisc_pkt_len_init netns=4026535338 mark=0x0 ifindex=13 proto=8 mtu=1500 len=74 10.0.1.45:54816->10.0.0.99:80(tcp)
0xffff97bfe0bb54e8      0           [curl]      netdev_core_pick_tx netns=4026535338 mark=0x0 ifindex=13 proto=8 mtu=1500 len=74 10.0.1.45:54816->10.0.0.99:80(tcp)
0xffff97bfe0bb54e8      0           [curl]        validate_xmit_skb netns=4026535338 mark=0x0 ifindex=13 proto=8 mtu=1500 len=74 10.0.1.45:54816->10.0.0.99:80(tcp)
0xffff97bfe0bb54e8      0           [curl]       netif_skb_features netns=4026535338 mark=0x0 ifindex=13 proto=8 mtu=1500 len=74 10.0.1.45:54816->10.0.0.99:80(tcp)
0xffff97bfe0bb54e8      0           [curl]  passthru_features_check netns=4026535338 mark=0x0 ifindex=13 proto=8 mtu=1500 len=74 10.0.1.45:54816->10.0.0.99:80(tcp)
0xffff97bfe0bb54e8      0           [curl]     skb_network_protocol netns=4026535338 mark=0x0 ifindex=13 proto=8 mtu=1500 len=74 10.0.1.45:54816->10.0.0.99:80(tcp)
0xffff97bfe0bb54e8      0           [curl]  skb_csum_hwoffload_help netns=4026535338 mark=0x0 ifindex=13 proto=8 mtu=1500 len=74 10.0.1.45:54816->10.0.0.99:80(tcp)
0xffff97bfe0bb54e8      0           [curl]       validate_xmit_xfrm netns=4026535338 mark=0x0 ifindex=13 proto=8 mtu=1500 len=74 10.0.1.45:54816->10.0.0.99:80(tcp)
0xffff97bfe0bb54e8      0           [curl]      dev_hard_start_xmit netns=4026535338 mark=0x0 ifindex=13 proto=8 mtu=1500 len=74 10.0.1.45:54816->10.0.0.99:80(tcp)
0xffff97bfe0bb54e8      0           [curl]         veth_xmit	[veth] netns=4026535338 mark=0x0 ifindex=13 proto=8 mtu=1500 len=74 10.0.1.45:54816->10.0.0.99:80(tcp)
0xffff97bfe0bb54e8      0           [curl]   skb_clone_tx_timestamp netns=4026535338 mark=0x0 ifindex=13 proto=8 mtu=1500 len=74 10.0.1.45:54816->10.0.0.99:80(tcp)
0xffff97bfe0bb54e8      0           [curl]        __dev_forward_skb netns=4026535338 mark=0x0 ifindex=13 proto=8 mtu=1500 len=74 10.0.1.45:54816->10.0.0.99:80(tcp)
0xffff97bfe0bb54e8      0           [curl]       __dev_forward_skb2 netns=4026535338 mark=0x0 ifindex=13 proto=8 mtu=1500 len=74 10.0.1.45:54816->10.0.0.99:80(tcp)
0xffff97bfe0bb54e8      0           [curl]         skb_scrub_packet netns=4026535338 mark=0x0 ifindex=13 proto=8 mtu=1500 len=74 10.0.1.45:54816->10.0.0.99:80(tcp)
0xffff97bfe0bb54e8      0           [curl]           eth_type_trans netns=4026535338 mark=0x0 ifindex=13 proto=8 mtu=1500 len=74 10.0.1.45:54816->10.0.0.99:80(tcp)
```

`eth_type_trans` is the last kernel function called, but it shouldn't be seen as the sign of lifetime termination. Searching `eth_type_trans` in kernel source, there are bunches of invocation not relevant to egress xmit at all.

`dev_hard_start_xmit` can be a better candidate, but there are other functions called after it, it's not suitable either.

In a word, I don't know how to stop tracking an egress skb.

**Comment by jschwinger233** [comment]
Maybe only handle `kfree_skbmem`?

Below is pwru output for an egress skb on bare metal:

```
0xffff97bfe0bb42e8      0           [curl]         __dev_queue_xmit netns=4026531840 mark=0x0 ifindex=2 proto=8 mtu=1500 len=74 192.168.0.110:43026->20.205.243.166:80(tcp)
0xffff97bfe0bb42e8      0           [curl]       qdisc_pkt_len_init netns=4026531840 mark=0x0 ifindex=2 proto=8 mtu=1500 len=74 192.168.0.110:43026->20.205.243.166:80(tcp)
0xffff97bfe0bb42e8      0           [curl]      netdev_core_pick_tx netns=4026531840 mark=0x0 ifindex=2 proto=8 mtu=1500 len=74 192.168.0.110:43026->20.205.243.166:80(tcp)
0xffff97bfe0bb42e8      0           [curl]          sch_direct_xmit netns=4026531840 mark=0x0 ifindex=2 proto=8 mtu=1500 len=74 192.168.0.110:43026->20.205.243.166:80(tcp)
0xffff97bfe0bb42e8      0           [curl]   validate_xmit_skb_list netns=4026531840 mark=0x0 ifindex=2 proto=8 mtu=1500 len=74 192.168.0.110:43026->20.205.243.166:80(tcp)
0xffff97bfe0bb42e8      0           [curl]        validate_xmit_skb netns=4026531840 mark=0x0 ifindex=2 proto=8 mtu=1500 len=74 192.168.0.110:43026->20.205.243.166:80(tcp)
0xffff97bfe0bb42e8      0           [curl]       netif_skb_features netns=4026531840 mark=0x0 ifindex=2 proto=8 mtu=1500 len=74 192.168.0.110:43026->20.205.243.166:80(tcp)
0xffff97bfe0bb42e8      0           [curl]  passthru_features_check netns=4026531840 mark=0x0 ifindex=2 proto=8 mtu=1500 len=74 192.168.0.110:43026->20.205.243.166:80(tcp)
0xffff97bfe0bb42e8      0           [curl]     skb_network_protocol netns=4026531840 mark=0x0 ifindex=2 proto=8 mtu=1500 len=74 192.168.0.110:43026->20.205.243.166:80(tcp)
0xffff97bfe0bb42e8      0           [curl]  skb_csum_hwoffload_help netns=4026531840 mark=0x0 ifindex=2 proto=8 mtu=1500 len=74 192.168.0.110:43026->20.205.243.166:80(tcp)
0xffff97bfe0bb42e8      0           [curl]       validate_xmit_xfrm netns=4026531840 mark=0x0 ifindex=2 proto=8 mtu=1500 len=74 192.168.0.110:43026->20.205.243.166:80(tcp)
0xffff97bfe0bb42e8      0           [curl]      dev_hard_start_xmit netns=4026531840 mark=0x0 ifindex=2 proto=8 mtu=1500 len=74 192.168.0.110:43026->20.205.243.166:80(tcp)
0xffff97bfe0bb42e8      0           [curl] e1000_xmit_frame	[e1000e] netns=4026531840 mark=0x0 ifindex=2 proto=8 mtu=1500 len=74 192.168.0.110:43026->20.205.243.166:80(tcp)
0xffff97bfe0bb42e8      0           [curl]       e1000_tso	[e1000e] netns=4026531840 mark=0x0 ifindex=2 proto=8 mtu=1500 len=74 192.168.0.110:43026->20.205.243.166:80(tcp)
0xffff97bfe0bb42e8      0           [curl]    e1000_tx_map	[e1000e] netns=4026531840 mark=0x0 ifindex=2 proto=8 mtu=1500 len=74 192.168.0.110:43026->20.205.243.166:80(tcp)
0xffff97bfe0bb42e8      0           [curl]   skb_clone_tx_timestamp netns=4026531840 mark=0x0 ifindex=2 proto=8 mtu=1500 len=74 192.168.0.110:43026->20.205.243.166:80(tcp)
0xffff97bfe0bb42e8     11        [<empty>]      __dev_kfree_skb_any netns=4026531840 mark=0x0 ifindex=2 proto=8 mtu=1500 len=74 192.168.0.110:43026->20.205.243.166:80(tcp)
0xffff97bfe0bb42e8     11        [<empty>]              consume_skb netns=4026531840 mark=0x0 ifindex=2 proto=8 mtu=1500 len=74 192.168.0.110:43026->20.205.243.166:80(tcp)
0xffff97bfe0bb42e8     11        [<empty>]   skb_release_head_state netns=4026531840 mark=0x0 ifindex=2 proto=8 mtu=1500 len=74 192.168.0.110:43026->20.205.243.166:80(tcp)
0xffff97bfe0bb42e8     11        [<empty>]                tcp_wfree netns=4026531840 mark=0x0 ifindex=2 proto=8 mtu=1500 len=74 192.168.0.110:43026->20.205.243.166:80(tcp)
0xffff97bfe0bb42e8     11        [<empty>]         skb_release_data netns=4026531840 mark=0x0 ifindex=2 proto=8 mtu=1500 len=74 192.168.0.110:43026->20.205.243.166:80(tcp)
0xffff97bfe0bb42e8     11        [<empty>]             kfree_skbmem netns=4026531840 mark=0x0 ifindex=2 proto=8 mtu=1500 len=74 192.168.0.110:43026->20.205.243.166:80(tcp)
```

As we see, there is a `kfree_skbmem` for "real" emission. 

The disadvantage of stopping tracking only at `kfree_skbmem` is, `pwru --filter-track-skb --filter-netns` will misfunction: when an skb goes away from the "wanted" netns, reaches a new netns we're not interested, it's expected to output nothing, but pwru continues displaying due to `--filter-track-skb`, the previously matched skb is still in the bpf map to skip the netns filter check.

**Comment by brb** [comment]
> Maybe only handle kfree_skbmem?

SGTM!

**Review by brb — APPROVED** [review]
:rocket:

**Pull request merged**